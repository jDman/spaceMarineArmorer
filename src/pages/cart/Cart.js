import React, { useEffect, useState, useCallback } from 'react';
import axios from 'axios';

import Button from '../../components/button/Button';
import Input from '../../components/form/input/Input';

import classes from './Cart.module.css';

const Cart = (props) => {
  const token = props.token;
  const [usersCartItems, setUsersCartItems] = useState([]);
  const [usersCartItemsNumber, setUsersCartItemsNumber] = useState(0);
  const [canOrder, setCanOrder] = useState(false);

  const getCart = useCallback(() => {
    return axios
      .get('http://localhost:4000/shop/armor/cart/items')
      .then((result) => {
        const inputConfigCartItems = result.data.items.map((item) => {
          return {
            armor: item.armor,
            itemId: item._id,
            type: item.armor.type,
            cost: item.armor.cost,
            totalCost: item.armor.cost * item.quantity,
            config: {
              elementType: 'input',
              elementConfig: {
                type: 'number',
                name: 'quantity',
                min: 0,
                max: 99,
                step: 1,
              },
              value: item.quantity,
              valid: true,
              touched: false,
            },
          };
        });

        const inputConfigCartItemsLength = inputConfigCartItems.length;

        setUsersCartItems(inputConfigCartItems);
        setUsersCartItemsNumber(inputConfigCartItemsLength);

        inputConfigCartItemsLength ? setCanOrder(true) : setCanOrder(false);
      })
      .catch((err) => console.log(err));
  }, []);

  const deleteCartItem = async (itemId) => {
    await props
      .deleteCartItem(itemId)
      .then(() => {
        getCart();
      })
      .catch((err) => console.error(err));
  };

  const inputChangedHandler = (newQuantity, itemId) => {
    const updatedUsersCartItems = [...usersCartItems].map((item) => {
      if (item.id === itemId) {
        item.config.value = newQuantity;
      }

      return item;
    });

    setUsersCartItems(updatedUsersCartItems);
  };

  const orderItems = async (items) => {
    await props
      .orderItems(items)
      .then((result) => {
        getCart();
      })
      .catch((err) => console.error(err));
  };

  useEffect(() => {
    axios.defaults.headers.common['Authorization'] = 'Bearer ' + token;
    axios.defaults.headers.common['Content-Type'] = 'application/json';

    getCart();
  }, [getCart, token]);

  return (
    <section className={classes.Cart}>
      <ul className={classes.CartItemList}>
        {usersCartItems.map((item) => (
          <li key={item.armor._id} className={classes.CartItem}>
            <div>
              <h3 className={classes.CartItemDetail}>{item.type} Armor Type</h3>
              <p className={classes.CartItemDetail}>
                Total price: {item.totalCost}
              </p>
              <Input
                elementType={'number'}
                elementConfig={item.config.elementConfig}
                value={item.config.value}
                invalid={!item.config.valid}
                touched={item.config.touched}
                hasValidation={item.config.validation}
                changed={(event) =>
                  inputChangedHandler(event.target.value, item.armor._id)
                }
              />
              <Button
                buttonText="Delete"
                classes={classes.CartItemDeleteButton}
                clickHandler={() => deleteCartItem(item.itemId)}
              />
            </div>
          </li>
        ))}
      </ul>

      <p className={classes.CartItemsInCounter}>
        Items in cart: {usersCartItemsNumber}
      </p>

      <Button
        buttonText="Order Item(s)"
        classes={
          !canOrder
            ? [classes.CartOrderButton, classes.CartOrderButtonDisabled]
            : classes.CartOrderButton
        }
        clickHandler={() => orderItems(usersCartItems)}
        disabled={!canOrder}
      />
    </section>
  );
};

export default Cart;
