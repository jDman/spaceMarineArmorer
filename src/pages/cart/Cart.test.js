import React from 'react';
import renderer, { act } from 'react-test-renderer';

import Cart from './Cart';

describe('<Cart />', () => {
  beforeEach(() => {
    global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
  });

  afterEach(() => {
    global.fetch.mockClear();
    delete global.fetch;
  });

  it('should render Cart page with empty cart items array returned from server', async () => {
    const result = {
      data: {
        items: [],
      },
      json: function () {
        return this.data;
      },
    };

    const component = renderer.create(<Cart />);

    await act(async () => {
      await fetch.mockReturnValue(Promise.resolve(result));
    });

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render Cart page with cart items array returned from server', async () => {
    const result = {
      data: {
        items: [
          {
            _id: '5e9584a3e78821c2d76c8900',
            armor: {
              createdBy: {
                userId: '5e70dfb438cee83fd9e004fd',
                userName: 'Freddy',
              },
              stock: 5,
              shield: 50,
              discount: 30,
              _id: '5e941bab2c9afa63aa5cadb3',
              type: 'helmet',
              cost: 200,
              protection: 'low',
              quality: 'low',
              description: 'Low tier trash',
              company: 'starscape_systems',
              createdAt: '2020-04-13T07:58:35.164Z',
              updatedAt: '2020-04-13T07:58:35.164Z',
              __v: 0,
            },
            quantity: 3,
          },
        ],
      },
      json: function () {
        return this.data;
      },
    };

    const component = renderer.create(<Cart />);

    await act(async () => {
      await fetch.mockReturnValue(Promise.resolve(result));
    });

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
