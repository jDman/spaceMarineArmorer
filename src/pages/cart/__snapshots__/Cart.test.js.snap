// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> should render Cart page with cart items array returned from server 1`] = `
<section
  className="Cart"
>
  <ul
    className="CartItemList"
  >
    <li
      className="CartItem"
    >
      <div>
        <h3
          className="CartItemDetail"
        >
          helmet
           Armor Type
        </h3>
        <p
          className="CartItemDetail"
        >
          Total price: 
          600
        </p>
        <div
          className="Input"
        >
          <label
            className="Label"
          >
            quantity
          </label>
          <input
            className="InputElement"
            max={99}
            min={0}
            name="quantity"
            onChange={[Function]}
            step={1}
            type="number"
            value={3}
          />
        </div>
        <button
          className="CartItemDeleteButton"
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </div>
    </li>
  </ul>
  <p
    className="CartItemsInCounter"
  >
    Items in cart: 
    1
  </p>
  <button
    className="CartOrderButton"
    onClick={[Function]}
    type="button"
  >
    Order Item(s)
  </button>
</section>
`;

exports[`<Cart /> should render Cart page with empty cart items array returned from server 1`] = `
<section
  className="Cart"
>
  <ul
    className="CartItemList"
  />
  <p
    className="CartItemsInCounter"
  >
    Items in cart: 
    0
  </p>
  <button
    className="CartOrderButton CartOrderButtonDisabled"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Order Item(s)
  </button>
</section>
`;
